syntaxes/GDScript.tmLanguage.json

```diff
@@ -309,37 +335,41 @@
             "name": "keyword.operator.assignment.gdscript"
         },
         "control_flow": {
-            "match": "\\b(?i:if|elif|else|for|while|break|continue|pass|return|match|yield|await)\\b",
+            "match": "\\b(?:if|elif|else|for|while|break|continue|pass|return|match|yield|await)\\b",
             "name": "keyword.control.gdscript"
         },
         "keywords": {
-            "match": "\\b(?i:class|class_name|extends|is|onready|tool|static|export|as|void|enum|preload|assert|breakpoint|rpc|sync|remote|master|puppet|slave|remotesync|mastersync|puppetsync|trait|namespace)\\b",
+            "match": "\\b(?:class|class_name|extends|is|onready|tool|static|export|as|void|enum|preload|assert|breakpoint|rpc|sync|remote|master|puppet|slave|remotesync|mastersync|puppetsync|trait|namespace)\\b",
             "name": "keyword.language.gdscript"
         },
         "letter": {
-            "match": "\\b(?i:true|false|null)\\b",
+            "match": "\\b(?:true|false|null)\\b",
             "name": "constant.language.gdscript"
         },
         "numbers": {
             "patterns": [
                 {
-                    "match": "\\b(?i:0x\\h*)\\b",
+                    "match": "\\b(?:0b[01_]+)\\b",
+                    "name": "constant.numeric.integer.binary.gdscript"
+                },
+                {
+                    "match": "\\b(?:0x[0-9A-Fa-f_]+)\\b",
                     "name": "constant.numeric.integer.hexadecimal.gdscript"
                 },
                 {
-                    "match": "\\b(?i:(\\d+\\.\\d*(e[\\-\\+]?\\d+)?))\\b",
+                    "match": "\\b(?:([0-9_]+\\.[0-9_]*(e[\\-\\+]?[0-9_]+)?))\\b",
                     "name": "constant.numeric.float.gdscript"
                 },
                 {
-                    "match": "\\b(?i:(\\.\\d+(e[\\-\\+]?\\d+)?))\\b",
+                    "match": "\\b(?:(\\.[0-9_]+(e[\\-\\+]?[0-9_]+)?))\\b",
                     "name": "constant.numeric.float.gdscript"
                 },
                 {
-                    "match": "\\b(?i:(\\d+e[\\-\\+]?\\d+))\\b",
+                    "match": "\\b(?:([0-9_]+e[\\-\\+]?\\[0-9_]))\\b",
                     "name": "constant.numeric.float.gdscript"
                 },
                 {
-                    "match": "\\b\\d+\\b",
+                    "match": "\\b[0-9_]+\\b",
```

Comment:
It's not right to allow `_` to intermix with numbers unconditionally. Cases like `_0` are valid variable identifiers, and must not be colored like number literals. This regex need to be carefully thought out and tested, because adding the `_` separator feature is not that simple. You need to ensure:
- `_` cannot appear at the beginning, or right after `e` in floats or `x` in hex literals
- `_` should not appear at the end, but it seems Godot allows it; better confirm this properly to know how to go about this
- consecutive `__` is not allowed

Reply:
On this particular line it would probably be enough to use `"\\b\\d[0-9_]*\\b"` if you won't forbid trailing `_`. On float and hex literals it'd probably be similar.

Reply:
I'm not interested in completely validating numeric literals with the grammar. Invalid cases will be syntax errors and the LSP should mark them as such.

Reply:
That's fair for invalid cases that aren't valid for other rules with less priority. But I believe valid variable names like `_0`, `_0e0`, `_0x0` will be taken over by this rule and be incorrectly colored as numbers. At the bare minimum, these regexes gotta exclude the initial `_` by using something like `\d[0-9_]*` instead of `[0-9_]+`.

But if you don't mind me completely validating it, the ideal would be:
int: `"\\b\\d+(?:_\\d+)*\\b"`
hex: `"\\b0x\\h+(?:_\\h+)*\\b"`
bin: `"\\b0b[01]+(?:_[01]+)*\\b"`
float1: `"\\b\\d+(?:_\\d+)*\\.(?:\\d+(?:_\\d+)*)?(?:[eE][-+]?\\d+(?:_\\d+)*)?\\b"`
float2: `"\\.\\d+(?:_\\d+)*(?:[eE][-+]?\\d+(?:_\\d+)*)?\\b"`
float3: `"\\b\\d+(?:_\\d+)*[eE][-+]?\\d+(?:_\\d+)*\\b"`

This should also fix:
- `[eE]` exponent which can be uppercase too
- a typo where the `\\` part of `\\d+` on line 338->368 wasn't removed
- float2 starts with `.` (non-word character) so it must not be preceded by a `\b` word boundary

Sorry for writing this in a comment instead of making a PR or something.

Reply:
> ... other rules with less priority. But I believe valid variable names like _0, _0e0, _0x0 will be taken over by this rule ...

Ah, I'm with you now; I misunderstood the problem. 

I'll add some new test cases and then pop in your rules. Thanks for writing them up!